doc/users-guide/users-guide-ipsec.adoc

```diff
@@ -0,0 +1,410 @@
+appear as events of subtype `ODP_EVENT_PACKET_IPSEC` and the `flag` field in
+the associated `odp_ipsec_packet_result_t` indicates `inline_mode`.
+
+For outbound IPsec processing, the `odp_ipsec_out_inline()` API operates as
+a "fire and forget" API. A success return code from this call indicates that
+the packet will be encrypted and transmitted to the `odp_pktio_t` indicated
+in the `odp_ipsec_out_inline_param_t` specified at the time of the call without
+any further application involvement. Only if a problem arises will the packet
+be returned to the application with an `odp_ipsec_packet_result_t` indicating
+the nature of the problem.
+
+==== SA Lifetimes
+A fundamental principle of good security is that the keying material
+associated with sessions has a limited lifetime. In effect, keys grow "stale"
+over time or due to being used to encrypt too much data. The metrics used
+to limit effective SA lifetimes are:
+
+* Duration (time)
+* Usage (volume of traffic using the keys)
+
+Associated with each of these metrics are "soft" and "hard" limits. When a
+hard limit is reached, the SA is expired and cannot be used further. To support
+graceful transition to a replacement SA, soft limits are used. A soft limit is
+similar to a "low fuel" warning light on a car. It alerts the application that
+the SA is nearing the end of its useful life and should be renegotiated even
+as the SA continues to work normally.
+
+ODP support for SA limits is based on packet/byte counts. Applications that
+wish to use time-based SA limits may do so on their own using the timing
+facilities that ODP provides. However, since especially with inline IPsec
+processing, the application may not have explicit knowledge of the traffic
+volumes associated with a given SA, support for usage-based limits is
+integrated into ODP IPsec support.
+
+At `odp_ipsec_sa_create()` time, one of the fields in the
+`odp_ipsec_sa_param_t` struct is the `odp_ipsec_lifetime_t` sub-structure.
+This struct allows hard and/or soft limits to be specified in terms of total
+bytes encrypted/decrypted, total packet count, or both. A limit specification
+of 0 indicates no limit for that metric. If either is specified, the limit
+is triggered on whichever occurs first. Given the defined behavior of hard vs.
+soft limits, the soft limits, if used, should always be specified as lower
+than the hard limits. These should be sufficiently lower to enable adequate
+time to switch over to a replacement SA before the hard limit is reached.
+
+As noted, when an SA hard limit is reached the SA immediately enters the
+disabled state and attempts to use it further are failed with an
+`odp_ipsec_result_t` that indicates a hard expiration limit.  When the soft
+limit is reached, this results in an `ODP_EVENT_IPSEC_STATUS` event being sent
+to the application on the default queue associated with the SA that has
+reached the soft limit. This status event has an `odp_ipsec_status_id_t` of
```

Comment:
Only in OUT INLINE mode, if I remember the outcome of discussions correctly.

Reply:
So how is this indicated in lookaside mode? The whole point of ODP providing the limit support is so the application doesn't have to track byte/packet counts itself, so it's expected that soft limit overruns will happen as part of lookaside processing as well.

Reply:
In lookaside mode soft limit expiration is reported as `warn` part of `ipsec_op_status` packet metadata.

Reply:
OK, corrected in v5.